source/Commands.cpp

```diff
@@ -333,4 +332,59 @@ namespace Commands
   {
     switch_focus_down_or_right(wm::Container::horizontalTiling);
   }
+
+  void switch_workspace(bool next)
+  {
+    Server &server = Server::getInstance();
+
+    for (auto const &output : server.outputManager.getOutputs())
+    {
+      auto it = std::find_if(output.get()->workspaces.begin(), output.get()->workspaces.end(),
+                            [](auto &w) {
+                              return w.get() == Server::getInstance().outputManager.getActiveWorkspace();
+                            });
+      if ((next && it == output.get()->getWorkspaces().end() - 1) ||
+          (!next && it == output.get()->getWorkspaces().begin())) {
+            return ;
+      }
+      auto newActiveWorkspace = it + (next ? 1 : -1);
+      server.outputManager.setActiveWorkspace(newActiveWorkspace->get());
+    }
+  }
+
+  void new_workspace()
+  {
+    Server &server = Server::getInstance();
+
+    for (auto const &output : server.outputManager.getOutputs())
+    {
+      auto it = std::find_if(output.get()->workspaces.begin(), output.get()->workspaces.end(),
+                            [](auto &w) {
+                              return w.get() == Server::getInstance().outputManager.getActiveWorkspace();
+                            });
+      output.get()->getWorkspaces().insert(output.get()->getWorkspaces().begin() + (std::distance(output.get()->workspaces.begin(), it) + 1),
+                                          std::make_unique<Workspace>(*(output.get())));
+    }
+    server.outputManager.workspacesNumber++;
+    switch_workspace(true);
+  }
+
+  void close_workspace()
+  {
+    Server &server = Server::getInstance();
+
+    if (server.outputManager.workspacesNumber == 2)
+      return ;
+    for (auto const &output : server.outputManager.getOutputs())
+    {
+      auto it = std::find_if(output.get()->workspaces.begin(), output.get()->workspaces.end(),
+                            [](auto &w) {
+                              return w.get() == Server::getInstance().outputManager.getActiveWorkspace();
+                            });
+      auto newActiveWorkspace = it + (it ==  output.get()->getWorkspaces().begin() ? 1 : -1);
+      server.outputManager.setActiveWorkspace(newActiveWorkspace->get());
+      output.get()->getWorkspaces().erase(output.get()->getWorkspaces().begin() + (std::distance(output.get()->workspaces.begin(), it)));
```

Comment:
See above about `output.get()->getWorkspaces().begin() + (std::distance(output.get()->workspaces.begin(), it))`